src/System.Text.Json/src/System/Text/Json/ThrowHelper.cs

```diff
@@ -20,6 +20,142 @@ private static ArgumentException GetArgumentException(string message)
+        {
+            throw GetArgumentException(SR.SpecialNumberValuesNotSupported);
+        }
+
+        public static void ThrowArgumentException(ExceptionResource resource, int minimumSize = 0)
+        {
+            if (resource == ExceptionResource.FailedToGetLargerSpan)
+            {
+                throw GetArgumentException(SR.FailedToGetLargerSpan);
+            }
+            else
+            {
+                Debug.Assert(resource == ExceptionResource.FailedToGetMinimumSizeSpan);
+                throw GetArgumentException(SR.Format(SR.FailedToGetMinimumSizeSpan, minimumSize));
+            }
+        }
+
+        public static void ThrowArgumentException(ReadOnlySpan<byte> propertyName, ReadOnlySpan<byte> value)
+        {
+            if (propertyName.Length > JsonConstants.MaxTokenSize)
+            {
+                ThrowArgumentException(SR.Format(SR.PropertyNameTooLarge, propertyName.Length));
+            }
+            else
+            {
+                Debug.Assert(value.Length > JsonConstants.MaxTokenSize);
+                ThrowArgumentException(SR.Format(SR.ValueTooLarge, value.Length));
+            }
+        }
+
+        public static void ThrowArgumentException(ReadOnlySpan<byte> propertyName, ReadOnlySpan<char> value)
+        {
+            if (propertyName.Length > JsonConstants.MaxTokenSize)
+            {
+                ThrowArgumentException(SR.Format(SR.PropertyNameTooLarge, propertyName.Length));
+            }
+            else
+            {
+                Debug.Assert(value.Length > JsonConstants.MaxCharacterTokenSize);
+                ThrowArgumentException(SR.Format(SR.ValueTooLarge, value.Length));
+            }
+        }
+
+        public static void ThrowArgumentException(ReadOnlySpan<char> propertyName, ReadOnlySpan<byte> value)
+        {
+            if (propertyName.Length > JsonConstants.MaxCharacterTokenSize)
+            {
+                ThrowArgumentException(SR.Format(SR.PropertyNameTooLarge, propertyName.Length));
+            }
+            else
+            {
+                Debug.Assert(value.Length > JsonConstants.MaxTokenSize);
+                ThrowArgumentException(SR.Format(SR.ValueTooLarge, value.Length));
+            }
+        }
+
+        public static void ThrowArgumentException(ReadOnlySpan<char> propertyName, ReadOnlySpan<char> value)
+        {
+            if (propertyName.Length > JsonConstants.MaxCharacterTokenSize)
+            {
+                ThrowArgumentException(SR.Format(SR.PropertyNameTooLarge, propertyName.Length));
+            }
+            else
+            {
+                Debug.Assert(value.Length > JsonConstants.MaxCharacterTokenSize);
+                ThrowArgumentException(SR.Format(SR.ValueTooLarge, value.Length));
+            }
+        }
+
+        public static void ThrowJsonWriterOrArgumentException(ReadOnlySpan<byte> propertyName, int currentDepth)
+        {
```

Comment:
Should `GetJsonWriterOrArgumentException` return the exception to be thrown and the exception to be thrown in here?

Otherwise, this split does not make sense.

Reply:
> return the exception to be thrown

Given it is returning different exceptions, what would the return type of `GetJsonWriterOrArgumentException` be, in that case?


> Otherwise, this split does not make sense.
Gd point. I will move the throw into the root method and remove the Get method entirely.